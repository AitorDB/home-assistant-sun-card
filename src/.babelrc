{
  "presets": [
    "@babel/env",
    "@babel/typescript"
  ],
  "plugins": [
    ["@babel/plugin-proposal-decorators", { "decoratorsBeforeExport": true }],
    "@babel/proposal-class-properties"
  ]
}
