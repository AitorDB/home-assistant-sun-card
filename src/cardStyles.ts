import { css } from 'lit-element'

export default css`
  .sun-card {
    --sun-card-lines: #464646;

    /* Use theme colors, fallback to default */
    --sun-card-text-color: var(--primary-text-color, #fff)
    --sun-card-subtitle-color: var(--secondary-text-color, #fff)

    color: var(--sun-card-text-color);
    padding: 1rem;
    
    font-family: var(--primary-font-family)
  }

  .sun-card-body {
    padding-top: 0.5rem;
  }

  .sun-card.sun-card-light {
    --sun-card-lines: #ececec;
    --sun-card-text-color: var(--primary-text-color, #000);
    --sun-card-subtitle-color: var(--secondary-text-color, #828282);
  }

  .sun-card-header {
    display: flex;
    justify-content: space-between;
  }
  
  .sun-card-footer .sun-card-footer-row {
    display: flex;
    justify-content: space-around;
    padding-top: 1.5rem;
  }

  .sun-card-title {
    font-size: 1.5rem;
    font-weight: 500;
    padding-bottom: 2rem;
    margin: 0;
  }

  .sun-card-text-container {
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  .sun-card-header .sun-card-text-subtitle {
    font-size: 1.15rem;
    font-weight: 400;
    padding-bottom: 0.25rem;
    color: var(--sun-card-subtitle-color);
  }

  .sun-card-header .sun-card-text-time {
    font-size: 1.85rem;
    font-weight: 400;
  }

  .sun-card-footer .sun-card-text-subtitle {
    font-size: 1.25rem;
    font-weight: 400;
    padding-bottom: 0.5rem;
    color: var(--sun-card-subtitle-color);
  }

  .sun-card-footer .sun-card-text-time {
    font-size: 1.25rem;
    font-weight: 500;
  }

  .sun-card-text-time-period {
    font-size: 0.75rem;
  }
`
