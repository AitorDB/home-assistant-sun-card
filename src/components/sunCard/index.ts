export * from './SunCard'
export * from './SunCardContent'
export * from './SunCardFooter'
export * from './SunCardGraph'
export * from './SunCardHeader'
