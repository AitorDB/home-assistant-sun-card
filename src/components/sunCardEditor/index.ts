export * from './SunCardEditor'
export * from './SunCardEditorContent'
