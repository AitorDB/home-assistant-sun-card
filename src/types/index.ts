import { LovelaceCardConfig } from 'custom-card-helpers'
import { I18N } from '../utils/I18N'


export type TSunCardFields = {
  sunrise?: boolean
  sunset?: boolean

  dawn?: boolean
  noon?: boolean
  dusk?: boolean

  azimuth?: boolean
  elevation?: boolean
}

export interface ISunCardConfig extends LovelaceCardConfig {
  i18n?: I18N
  darkMode?: boolean
  language?: string
  
  title?: string,
  component?: string
  use12hourClock?: boolean

  fields?: TSunCardFields
}

export interface IConfigChangedEvent <T> extends CustomEvent<T> {
  target: CustomEvent['target'] & {
    configValue: string
    selected?: string
    checked?: boolean
  }
}

export type TSunInfo = {
  sunrise: number,
  sunset: number

  dawnProgressPercent: number
  dayProgressPercent: number
  duskProgressPercent: number
  
  sunAboveHorizon: boolean
  sunPercentOverHorizon: number
  sunPosition: {
    x: number
    y: number
  }
}

export enum ESunCardErrors {
  SunIntegrationNotFound = 'SunIntegrationNotFound'
}

export type TSunCardData = {
  azimuth: number
  elevation: number

  sunInfo: TSunInfo

  times: TSunCardTimes

  error?: ESunCardErrors
}

export type TSunCardTimes = {
  dawn: Date
  dusk: Date
  noon: Date
  sunrise: Date
  sunset: Date
}

export enum ESunCardI18NKeys {
  Azimuth = 'azimuth',
  Dawn = 'dawn',
  Dusk = 'dusk',
  Elevation = 'elevation',
  Noon = 'noon',
  Sunrise = 'sunrise',
  Sunset = 'sunset'
}

export type TSunCardI18NKeys = { [key in ESunCardI18NKeys ]: string } | { errors: TSunCardI18NErrorKeys }

export type TSunCardI18N = Record<string, unknown>

export type TSunCardI18NErrorKeys = {
  [key in ESunCardErrors]: string
}
