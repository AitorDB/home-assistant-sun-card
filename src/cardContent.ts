import { html, TemplateResult } from 'lit-element'
import { TSunCardConfig, TSunCardData, TSunCardTexts, TSunCardTime } from './types'

export class SunCardContent {
  static generate (data: TSunCardData, localization: TSunCardTexts, config: TSunCardConfig): TemplateResult {
    return html`
      <ha-card>
        <div class="sun-card ${config.darkMode ? '' : 'sun-card-light'}">
          ${this.generateHeader(data, localization, config)}
          ${this.generateBody(data)}
          ${this.generateFooter(data, localization, config)}
        </div>
      </ha-card>
    `
  }

  private static generateHeader (data: TSunCardData, localization: TSunCardTexts, config: TSunCardConfig): TemplateResult {
    const title = config.title !== undefined ? html`
      <h1 class="sun-card-title">${config.title}</h1>
    ` : html``

    return html`
      ${title}
      <div class="sun-card-header">
        <div class="sun-card-text-container">
          <span class="sun-card-text-subtitle">${localization.Sunrise}</span>
          ${data?.times.sunrise ? this.generateTime(data.times.sunrise) : ''}

        </div>
        <div class="sun-card-text-container">
          <span class="sun-card-text-subtitle">${localization.Sunset}</span>
          ${data?.times.sunset ? this.generateTime(data.times.dawn) : ''}
        </div>
      </div>
    `
  }

  private static generateBody (data: TSunCardData): TemplateResult {
    const sunID = Math.random().toString(36).replace('0.', '')
    const dawnID = Math.random().toString(36).replace('0.', '')
    const dayID = Math.random().toString(36).replace('0.', '')
    const duskID = Math.random().toString(36).replace('0.', '')

    return html`
      <div class="sun-card-body">
        <svg viewBox="0 0 550 150" xmlns="http://www.w3.org/2000/svg">
          <defs>
            <linearGradient id="${sunID}" x1="0%" y1="0%" x2="0%" y2="100%">
              <stop offset="0%" style="stop-color:#f9d05e;stop-opacity:1" />
              <stop offset="${data?.sunPercentOverHorizon ?? 0}%" style="stop-color:#f9d05e;stop-opacity:1" />
              <stop offset="${data?.sunPercentOverHorizon ?? 0}%" style="stop-color:rgb(0,0,0,0);stop-opacity:1" />
            </linearGradient>
            
            <linearGradient id="${dawnID}" x1="0%" y1="0%" x2="100%" y2="0%">
              <stop offset="0%" style="stop-color:#393b78;stop-opacity:1" />
              <stop offset="${data?.dawnProgressPercent ?? 0}%" style="stop-color:#393b78;stop-opacity:1" />
              <stop offset="${data?.dawnProgressPercent ?? 0}%" style="stop-color:rgb(0,0,0,0);stop-opacity:1" />
            </linearGradient>
            
            <linearGradient id="${dayID}" x1="0%" y1="0%" x2="100%" y2="0%">
              <stop offset="0%" style="stop-color:#8ebeeb;stop-opacity:1" />
              <stop offset="${data?.dayProgressPercent ?? 0}%" style="stop-color:#8ebeeb;stop-opacity:1" />
              <stop offset="${data?.dayProgressPercent ?? 0}%" style="stop-color:rgb(0,0,0,0);stop-opacity:1" />
            </linearGradient>
            
            <linearGradient id="${duskID}" x1="0%" y1="0%" x2="100%" y2="0%">
              <stop offset="0%" style="stop-color:#393b78;stop-opacity:1" />
              <stop offset="${data?.duskProgressPercent ?? 0}%" style="stop-color:#393b78;stop-opacity:1" />
              <stop offset="${data?.duskProgressPercent ?? 0}%" style="stop-color:rgb(0,0,0,0);stop-opacity:1" />
            </linearGradient>
          </defs>
          <path class="sun-card-sun-line" d="M5,146 C29,153 73,128 101,108 C276,-29 342,23 449,108 C473,123 509,150 545,146" fill="none" stroke="var(--sun-card-lines)" shape-rendering="geometricPrecision" />
          <path d="M5,146 C29,153 73,128 101,108 L 5 108" fill="url(#${dawnID})" opacity="${data?.dawnProgressPercent ? 1 : 0}" stroke="url(#${dawnID})" shape-rendering="geometricPrecision" />
          <path d="M101,108 C276,-29 342,23 449,108 L 104,108" fill="url(#${dayID})" opacity="${data?.dayProgressPercent ? 1 : 0}" stroke="url(#${dayID})" shape-rendering="geometricPrecision" />
          <path d="M449,108 C473,123 509,150 545,146 L 545 108" fill="url(#${duskID})" opacity="${data?.duskProgressPercent ? 1 : 0}" stroke="url(#${duskID})" shape-rendering="geometricPrecision" />
          <line x1="5" y1="108" x2="545" y2="108" stroke="var(--sun-card-lines)" />
          <line x1="101" y1="25" x2="101" y2="100" stroke="var(--sun-card-lines)" />
          <line x1="449" y1="25" x2="449" y2="100" stroke="var(--sun-card-lines)" />
          <circle cx="${data?.sunPosition.x ?? 0}" cy="${data?.sunPosition.y ?? 0}" r="17" opacity="${data?.sunPercentOverHorizon ? 1 : 0}" stroke="none" fill="url(#${sunID})" shape-rendering="geometricPrecision" />
        </svg>
      </div>
    `
  }

  private static generateFooter (data: TSunCardData, localization: TSunCardTexts, config: TSunCardConfig): TemplateResult {
    const upperRow = html`
      <div class="sun-card-footer-row">
        <div class="sun-card-text-container">
          <span class="sun-card-text-subtitle">${localization.Dawn}</span>
          ${data?.times.dawn ? this.generateTime(data.times.dawn) : ''}
        </div>
        <div class="sun-card-text-container">
          <span class="sun-card-text-subtitle">${localization.Noon}</span>
          ${data?.times.noon ? this.generateTime(data.times.noon) : ''}
        </div>
        <div class="sun-card-text-container">
          <span class="sun-card-text-subtitle">${localization.Dusk}</span>
          ${data?.times.dusk ? this.generateTime(data.times.dusk) : ''}
        </div>
      </div>
    `

    let bottomRow = html``
    if (config.showAzimuth || config.showElevation) {
      const azimuth = config.showAzimuth ? html`
        <div class="sun-card-text-container">
          <span class="sun-card-text-subtitle">${localization.Azimuth}</span>
          <span class="sun-card-dawn-time sun-card-text-time">${data?.azimuth ?? ''}</span>
        </div>
      ` : html``

      const elevation = config.showElevation ? html`
        <div class="sun-card-text-container">
          <span class="sun-card-text-subtitle">${localization.Elevation}</span>
          <span class="sun-card-dawn-time sun-card-text-time">${data?.elevation ?? ''}</span>
        </div>
      ` : html``
  
      bottomRow = html`
        <div class="sun-card-footer-row">
          ${azimuth}
          ${elevation}
        </div>
      `
    }

    return html`
      <div class="sun-card-footer">
        ${upperRow}
        ${bottomRow}
      </div>
    `
  }

  private static generateTime (time: TSunCardTime) {
    if (time.period) {
      return html`
        <span class="sun-card-text-time">
          ${time.time} <span class="sun-card-text-time-period">${time.period}</span>
        </span>
      `
    }
    
    return html`
      <span class="sun-card-text-time">${time.time}</span>
    `
  }
}
