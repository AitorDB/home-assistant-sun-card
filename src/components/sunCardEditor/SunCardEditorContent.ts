import { html,TemplateResult } from 'lit-html'

import { Constants } from '../../constants'
import { ESunCardI18NKeys,IConfigChangedEvent, ISunCardConfig } from '../../types'
import { EventUtils } from '../../utils/EventUtils'

export type TSunCardEditorContentEvents = {
  configChanged: IConfigChangedEvent<{ value: unknown }>
}

export class SunCardEditorContent extends EventUtils<TSunCardEditorContentEvents> {
  config: ISunCardConfig

  constructor (config: ISunCardConfig) {
    super()
    this.config = config
  }

  public render (): TemplateResult {
    return html`
      <div class="card-config">
        <div>
          ${this.renderTitleEditor()}
        </div>
        <div>
          ${this.renderLanguageEditor()}
        </div>
        <div>
          ${this.renderDarkModeEditor()}
        </div>
        <div>
          ${this.render12HourClockEditor()}
        </div>
        <div>
          ${this.renderFieldsEditor()}
        </div>
      </div>
    `
  }

  private onConfigChanged (event: TSunCardEditorContentEvents['configChanged']) {
    this.emit('configChanged', event)
  }

  private renderTitleEditor (): TemplateResult {
    return html`
      <paper-input
        label="Title (Optional)"
        .configValue=${'title'}
        .value=${this.config?.title ?? ''}
        @value-changed=${(event) => this.onConfigChanged(event)}
      >
      </paper-input>
    `
  }

  private renderLanguageEditor (): TemplateResult {
    // TODO: Add language full name
    const selectedLanguage = Object.keys(Constants.LOCALIZATION_LANGUAGES).indexOf(this.config?.language ?? '') + 1

    return html`
      <paper-dropdown-menu
        label="Language"
        .configValue=${'language'}
        @value-changed=${(event) => this.onConfigChanged(event)}
      >
        <paper-listbox slot="dropdown-content" selected="${selectedLanguage}">
          <paper-item label="default">Default</paper-item>
          ${Object.keys(Constants.LOCALIZATION_LANGUAGES).map((language) => html`
            <paper-item label="${language}">${language}</paper-item>
          `)}
        </paper-listbox>
      </paper-dropdown-menu>
    `
  }

  private renderDarkModeEditor (): TemplateResult {
    const selectedDarkMode = this.config?.darkMode ?? 'default'
    return html`
      <label id="theme">Theme:</label>
      <paper-radio-group
        aria-labelledby="theme"
        .configValue=${'darkMode'}
        .selected=${selectedDarkMode.toString()}
        @paper-radio-group-changed=${(event) => this.onConfigChanged(event)}
      >
        <paper-radio-button name="default">Default</paper-radio-button>
        <paper-radio-button name="true">Dark</paper-radio-button>
        <paper-radio-button name="false">Light</paper-radio-button>
      </paper-radio-group>
    `
  }

  private render12HourClockEditor (): TemplateResult {
    const selectedClockMode = this.config?.use12hourClock ?? 'default'

    return html`
      <label id="clock">Clock mode:</label>
      <paper-radio-group
        aria-labelledby="clock"
        .configValue=${'use12hourClock'}
        .selected=${selectedClockMode.toString()}
        @paper-radio-group-changed=${(event) => this.onConfigChanged(event)}
      >
        <paper-radio-button name="default">Default</paper-radio-button>
        <paper-radio-button name="true">12 hours</paper-radio-button>
        <paper-radio-button name="false">24 hourse</paper-radio-button>
      </paper-radio-group>
    `
  }

  private renderFieldsEditor (): TemplateResult {
    return html`
      <label>Card fields:</label>
      <ul>
        ${Object.entries(ESunCardI18NKeys).map(([name, configValue]) => {
    return html`
            <li><ha-switch .configValue=${configValue} .checked=${this.config.fields?.[configValue] ?? Constants.DEFAULT_CONFIG.fields![configValue]} @change=${(event) => this.onConfigChanged(event)}></ha-switch> ${name}</li>
          `
  })}
      </ul>
    `
  }
}
