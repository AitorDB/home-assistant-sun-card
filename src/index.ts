import { HomeAssistant } from "custom-card-helpers";
import { customElement, LitElement, state } from "lit-element";

import cardStyles from "./cardStyles";
import { Constants } from "./constants";
import { SunCardContent } from "./cardContent";
import { TSunCardConfig, TSunCardData, TSunCardTime } from "./types";

@customElement("sun-card")
class SunCard extends LitElement {
  static readonly cardType = "sun-card";
  static readonly cardName = "Sun Card";
  static readonly cardDescription =
    "Custom card that display a graph to track the sun position and related events";

  @state()
  private config: TSunCardConfig = {};

  @state()
  private data!: TSunCardData;

  private hasRendered = false;
  private lastHass!: HomeAssistant;

  set hass(hass: HomeAssistant) {
    this.lastHass = hass;

    if (!this.hasRendered) {
      return;
    }

    this.processLastHass();
  }

  calculatePositionAndProgressesByTime(hass: HomeAssistant) {
    const sunLine = this.shadowRoot?.querySelector("path") as SVGPathElement;
    const sunrise = new Date(hass.states["sun.sun"].attributes.next_rising);
    const sunset = new Date(hass.states["sun.sun"].attributes.next_setting);
    const eventsAt = {
      dayStart: 0,
      sunrise: this.convertDateToMinutesSinceDayStarted(sunrise),
      sunset: this.convertDateToMinutesSinceDayStarted(sunset),
      dayEnd: 23 * 60 + 59,
    };

    const now = new Date();
    const minutesSinceTodayStarted =
      this.convertDateToMinutesSinceDayStarted(now);

    // Dawn section position [0 - 105]
    const dawnSectionPosition =
      (Math.min(minutesSinceTodayStarted, eventsAt.sunrise) * 105) /
      eventsAt.sunrise;

    // Day section position [106 - 499]
    const minutesSinceDayStarted = Math.max(
      minutesSinceTodayStarted - eventsAt.sunrise,
      0
    );
    const daySectionPosition =
      (Math.min(minutesSinceDayStarted, eventsAt.sunset - eventsAt.sunrise) *
        (499 - 106)) /
      (eventsAt.sunset - eventsAt.sunrise);

    // Dusk section position [500 - 605]
    const minutesSinceDuskStarted = Math.max(
      minutesSinceTodayStarted - eventsAt.sunset,
      0
    );
    const duskSectionPosition =
      (minutesSinceDuskStarted * (605 - 500)) /
      (eventsAt.dayEnd - eventsAt.sunset);

    const position =
      dawnSectionPosition + daySectionPosition + duskSectionPosition;
    const sunPosition = sunLine.getPointAtLength(position);

    const dawnProgressPercent =
      (100 * (sunPosition.x - Constants.EVENT_X_POSITIONS.dayStart)) /
      (Constants.EVENT_X_POSITIONS.sunrise -
        Constants.EVENT_X_POSITIONS.dayStart);
    const dayProgressPercent =
      (100 * (sunPosition.x - Constants.EVENT_X_POSITIONS.sunrise)) /
      (Constants.EVENT_X_POSITIONS.sunset -
        Constants.EVENT_X_POSITIONS.sunrise);
    const duskProgressPercent =
      (100 * (sunPosition.x - Constants.EVENT_X_POSITIONS.sunset)) /
      (Constants.EVENT_X_POSITIONS.dayEnd - Constants.EVENT_X_POSITIONS.sunset);

    const sunYTop = sunPosition.y - Constants.SUN_RADIUS;
    const yOver = Constants.HORIZON_Y - sunYTop;
    let sunPercentOverHorizon = 0;
    if (yOver > 0) {
      sunPercentOverHorizon = Math.min(
        (100 * yOver) / (2 * Constants.SUN_RADIUS),
        100
      );
    }

    return {
      dawnProgressPercent,
      dayProgressPercent,
      duskProgressPercent,
      sunPercentOverHorizon,
      sunPosition: { x: sunPosition.x, y: sunPosition.y },
    };
  }

  convertDateToMinutesSinceDayStarted(date: Date) {
    return date.getHours() * 60 + date.getMinutes();
  }

  parseTime(timeText: string, locale?: string) {
    const regex = /\d{1,2}[:.]\d{1,2}|[AMP]+/g;
    const date = new Date(timeText);
    const { language, timeFormat } = this.getConfig();
    const result = date
      .toLocaleTimeString(locale ?? language, { hour12: timeFormat === "12h" })
      .match(regex) as [string, ("AM" | "PM")?];

    if (!result && !locale) {
      return this.parseTime(timeText, Constants.DEFAULT_CONFIG.language);
    }

    const [time, period] = result;
    return { time, period };
  }
  calculateTimeBetweenDates(date1: Date, date2: Date): TSunCardTime {
    const timeMs = Math.abs(date1 - date2);
    const { language } = this.getConfig();
    const time = new Date(timeMs).toLocaleTimeString(language, {
      hour12: false,
    });
    return { time };
  }
  processLastHass() {
    if (!this.lastHass) {
      return;
    }

    this.config.darkMode =
      this.config.darkMode ?? this.lastHass.themes.darkMode;
    this.config.language =
      this.config.language ??
      this.lastHass.locale?.language ??
      this.lastHass.language;
    this.config.timeFormat =
      this.config.timeFormat ??
      this.getTimeFormatByLanguage(this.config.language);

    const times = {
      dawn: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_dawn
      ),
      dusk: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_dusk
      ),
      noon: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_noon
      ),
      sunrise: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_rising
      ),
      sunset: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_setting
      ),
    };

    const {
      dawnProgressPercent,
      dayProgressPercent,
      duskProgressPercent,
      sunPercentOverHorizon,
      sunPosition,
    } = this.calculatePositionAndProgressesByTime(this.lastHass);

    const lengthOfDay = this.calculateTimeBetweenDates(
      new Date(this.lastHass.states["sun.sun"].attributes.next_rising),
      new Date(this.lastHass.states["sun.sun"].attributes.next_setting)
    );

    const daylightLeft = this.calculateTimeBetweenDates(
      new Date(this.lastHass.states["sun.sun"].attributes.next_setting),
      new Date()
    );

    const data: TSunCardData = {
      azimuth: this.lastHass.states["sun.sun"].attributes.azimuth,
      dawnProgressPercent,
      dayProgressPercent,
      duskProgressPercent,
      elevation: this.lastHass.states["sun.sun"].attributes.elevation,
      sunPercentOverHorizon,
      sunPosition,
      times,
      lengthOfDay,
      daylightLeft,
    };

    this.data = data;
  }

  getConfig() {
    const config: TSunCardConfig = {};
    config.darkMode = this.config.darkMode ?? Constants.DEFAULT_CONFIG.darkMode;
    config.language = this.config.language ?? Constants.DEFAULT_CONFIG.language;
    config.showAzimuth =
      this.config.showAzimuth ?? Constants.DEFAULT_CONFIG.showAzimuth;
    config.showElevation =
      this.config.showElevation ?? Constants.DEFAULT_CONFIG.showElevation;
    config.timeFormat =
      this.config.timeFormat ?? Constants.DEFAULT_CONFIG.timeFormat;
    config.title = this.config.title;

    if (
      !Object.keys(Constants.LOCALIZATION_LANGUAGES).includes(config.language!)
    ) {
      config.language = Constants.DEFAULT_CONFIG.language;
    }

    return config;
  }

  getTimeFormatByLanguage(language: string) {
    const date = new Date();
    const time = date.toLocaleTimeString(language).toLocaleLowerCase();
    return time.includes("pm") || time.includes("am") ? "12h" : "24h";
  }

  setConfig(config: TSunCardConfig) {
    this.config = { ...config };
  }

  protected render() {
    const config = this.getConfig();
    const language = config.language!;
    const localization = Constants.LOCALIZATION_LANGUAGES[language];
    return SunCardContent.generate(this.data, localization, config);
  }

  protected updated(changedProperties) {
    super.updated(changedProperties);

    if (!this.hasRendered) {
      this.hasRendered = true;
      this.processLastHass();
    }
  }

  static get styles() {
    return cardStyles;
  }
}

window.customCards = window.customCards || [];
window.customCards.push({
  type: SunCard.cardType,
  name: SunCard.cardName,
  description: SunCard.cardDescription,
});
