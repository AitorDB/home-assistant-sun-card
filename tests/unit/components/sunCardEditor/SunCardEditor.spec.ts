import { HomeAssistant } from 'custom-card-helpers'
import { css, CSSResult } from 'lit'

import { SunCardEditor, TSunCardEditorContentEvents } from '../../../../src/components/sunCardEditor'
import { ISunCardConfig } from '../../../../src/types'
import { CustomSnapshotSerializer, TemplateResultTestHelper } from '../../../helpers/TestHelpers'
import { SunCardEditorContent } from '../../../mocks/SunCardEditorContent'

jest.mock('../../../../src/components/SunErrorContent', () => require('../../../mocks/SunErrorContent'))
jest.mock('../../../../src/utils/HelperFunctions', () => require('../../../mocks/HelperFunctions'))
jest.mock('../../../../src/utils/I18N', () => require('../../../mocks/I18N'))
jest.mock('../../../../src/cardStyles', () => css``)
jest.mock('../../../../src/components/sunCardEditor/SunCardEditorContent.ts', () => require('../../../mocks/SunCardEditorContent'))

expect.addSnapshotSerializer(new CustomSnapshotSerializer())

describe('SunCardEditor', () => {
  let sunCardEditor: SunCardEditor

  beforeEach(() => {
    SunCardEditorContent.onMock = jest.fn()
    sunCardEditor = new SunCardEditor()
  })

  describe('hass', () => {
    it('updates hass property with provided value', () => {
      const expectedValue = { language: 'test' } as HomeAssistant
      sunCardEditor.hass = expectedValue
      expect(sunCardEditor.hass).toEqual(expectedValue)
    })
  })

  describe('setConfig', () => {
    it('updates config property with provided value', () => {
      const expectedValue = { language: 'test' } as ISunCardConfig
      sunCardEditor.setConfig(expectedValue)
      expect(sunCardEditor['config']).toEqual(expectedValue)
    })
  })

  describe('configChanged', () => {
    let dispatchEventMock
    beforeEach(() => {
      dispatchEventMock = jest.fn()
      sunCardEditor.dispatchEvent = dispatchEventMock
    })

    it('retrieves correctly the value from input', () => {
      const config = { target: { configValue: 'test' }, detail: { value: 'test' } } as TSunCardEditorContentEvents['configChanged']
      sunCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config).toEqual(expect.objectContaining({
        test: 'test'
      }))
    })

    it('retrieves correctly the value from select', () => {
      const config = { target: { configValue: 'test', selected: 'true' } } as TSunCardEditorContentEvents['configChanged']
      sunCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config).toEqual(expect.objectContaining({
        test: true
      }))
    })

    it('retrieves correctly the value from checkbox', () => {
      const config = { target: { configValue: 'test', checked: true } } as TSunCardEditorContentEvents['configChanged']
      sunCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config).toEqual(expect.objectContaining({
        test: true
      }))
    })

    it('deletes the config property when it the value is default', () => {
      const config = { target: { configValue: 'test' }, detail: { value: 'default' } } as TSunCardEditorContentEvents['configChanged']
      sunCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config.test).toBeUndefined()
    })

    it('deletes the config property when it the value is undefined', () => {
      const config = { target: { configValue: 'test' } } as TSunCardEditorContentEvents['configChanged']
      sunCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config.test).toBeUndefined()
    })

    it('deletes the config property when it the value is an empty string', () => {
      const config = { target: { configValue: 'test' }, detail: { value: '' } } as TSunCardEditorContentEvents['configChanged']
      sunCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config.test).toBeUndefined()
    })

    it('transform "true"/"false" to boolean', () => {
      const config = { target: { configValue: 'test' }, detail: { value: 'false' } } as TSunCardEditorContentEvents['configChanged']
      sunCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config).toEqual(expect.objectContaining({
        test: false
      }))
    })

    it('handles field configuration correctly', () => {
      const config = { target: { configValue: 'dawn' }, detail: { value: true } } as TSunCardEditorContentEvents['configChanged']
      sunCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config).toEqual(expect.objectContaining({
        fields: {
          dawn: true
        }
      }))
    })

    it('dispatches an event', () => {
      const config = { target: { configValue: 'test' }, detail: { value: 'test' } } as TSunCardEditorContentEvents['configChanged']
      sunCardEditor.configChanged(config)

      expect(sunCardEditor.dispatchEvent).toHaveBeenCalledTimes(1)
    })
  })

  describe('render', () => {
    it('starts to listen then configChanged event from SunCardEditorContent', () => {
      sunCardEditor['render']()
      expect(SunCardEditorContent.onMock).toHaveBeenCalledWith('configChanged', expect.any(Function))
    })

    it('renders the sun card editor content render result', async () => {
      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof sunCardEditor['render']>
      element.templateResultFunction = () => sunCardEditor['render']()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })
  })

  describe('get styles', () => {
    it('returns a CSSResult', () => {
      expect(SunCardEditor.styles).toBeInstanceOf(CSSResult)
    })
  })
})
