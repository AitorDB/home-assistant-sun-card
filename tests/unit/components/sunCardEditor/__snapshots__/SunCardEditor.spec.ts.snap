// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SunCardEditor render renders the sun card editor content render result 1`] = `
<!---->
      <div class="card-config">
        SUN CARD EDITOR CONTENT
      </div>
    
`;
